### ISSUE TITLE
Readonly<Float32Array> not assignable to Float32Array

### ISSUE THREAD

**Issue by chrisui** [microsoft/TypeScript#31253]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**

**Code**

```ts
function update(b: Readonly<Float32Array>) {
    const c = copy(b);
    add(c, c);
}

function add(a: Float32Array, b: Float32Array, c: Float32Array = a) {
    c[0] = a[0] + b[0];
}

function copy(a: Float32Array) {
    return new Float32Array(a);
}
```

**Expected behavior:**

No error.

_Though I would expect not to be able to pass a Readonly value to a function which can mutate it but that's [another issue](https://github.com/Microsoft/TypeScript/issues/13347)_.

**Actual behavior:**

Error on line 2:

```
Argument of type 'Readonly<Float32Array>' is not assignable to parameter of type 'Float32Array'.
  Type 'Readonly<Float32Array>' is missing the following properties from type 'Float32Array': [Symbol.iterator], [Symbol.toStringTag]
```

**Playground Link:** 

[See in playground](https://www.typescriptlang.org/play/index.html#src=function%20update(b%3A%20Readonly%3CFloat32Array%3E)%20%7B%0D%0A%20%20%20%20const%20c%20%3D%20copy(b)%3B%0D%0A%20%20%20%20add(c%2C%20c)%3B%0D%0A%7D%0D%0A%0D%0Afunction%20add(a%3A%20Float32Array%2C%20b%3A%20Float32Array%2C%20c%3A%20Float32Array%20%3D%20a)%20%7B%0D%0A%20%20%20%20c%5B0%5D%20%3D%20a%5B0%5D%20%2B%20b%5B0%5D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20copy(a%3A%20Float32Array)%20%7B%0D%0A%20%20%20%20return%20new%20Float32Array(a)%3B%0D%0A%7D)

Note I can workaround the issue by using mappable types on the function arguments but this is pretty verbose and not something I will want to litter a codebase with - Also not going to work with third-party libraries. See that [here](https://www.typescriptlang.org/play/index.html#src=function%20update(b%3A%20Readonly%3CFloat32Array%3E)%20%7B%0D%0A%20%20%20%20const%20c%20%3D%20copy(b)%3B%0D%0A%20%20%20%20add(c%2C%20c)%3B%0D%0A%7D%0D%0A%0D%0Afunction%20add(a%3A%20Mutable%3CFloat32Array%3E%2C%20b%3A%20Mutable%3CFloat32Array%3E%2C%20c%3A%20Mutable%3CFloat32Array%3E%20%3D%20a)%20%7B%0D%0A%20%20%20%20c%5B0%5D%20%3D%20a%5B0%5D%20%2B%20b%5B0%5D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20copy(a%3A%20Readonly%3CFloat32Array%3E)%20%7B%0D%0A%20%20%20%20return%20new%20Float32Array(a)%3B%0D%0A%7D%0D%0A%0D%0Atype%20Mutable%3CT%3E%20%3D%20%7B%0D%0A%20%20%20%20-readonly%20%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%3B%0D%0A%7D%3B). Edit: Actually seen another way to do it which is much less verbose [here](https://www.typescriptlang.org/play/index.html#src=function%20update(b%3A%20Readonly%3CFloat32Array%3E)%20%7B%0D%0A%20%20%20%20const%20c%20%3D%20copy(b%20as%20Float32Array)%3B%0D%0A%20%20%20%20add(c%2C%20c)%3B%0D%0A%7D%0D%0A%0D%0Afunction%20add(a%3A%20Float32Array%2C%20b%3A%20Float32Array%2C%20c%3A%20Float32Array%20%3D%20a)%20%7B%0D%0A%20%20%20%20c%5B0%5D%20%3D%20a%5B0%5D%20%2B%20b%5B0%5D%3B%0D%0A%7D%0D%0A%0D%0Afunction%20copy(a%3A%20Float32Array)%20%7B%0D%0A%20%20%20%20return%20new%20Float32Array(a)%3B%0D%0A%7D). _This however loops again back to [this issue](https://github.com/Microsoft/TypeScript/issues/13347) since it really should be raising an error (if in strict mode)._

**Related Issues:** 

N/A

**Comment by ExE-Boss** [comment]
This is caused by the fact that symbols aren’t preserved by mapped types.

**Comment by andrewbranch** [comment]
> This is caused by the fact that symbols aren’t preserved by mapped types.

Yep, or more specifically, they’re not exposed by `keyof`. I’m trying to dig up why that is, but changing it would likely be a _huge_ breaking change... @weswigham what are your thoughts here?

**Comment by weswigham** [comment]
Certain builtin symbols we know of predate `unique symbol`s and do not appear in any object key lists - #24738 was originally opened to address this, and `typesVersions` was made to make the migration to real `unique symbol`s for these global symbols better.

**Comment by andrewbranch** [comment]
Got it. In a sense, then, this is a duplicate of #24622. At least, I think it’s blocked on #24738.

**Comment by Jamesernator** [comment]
> Certain builtin symbols (ie, Symbol.iterator) we know of predate unique symbols and do not appear in any object key lists

Glad to see this is being fixed, just found out that this does not currently work:

```ts
const asyncIterable: AsyncIterable<number> = Object.freeze({
  * [Symbol.asyncIterator]() {
    yield 1
    yield 2
  },
})
```

**Comment by andrewbranch** [comment]
@DanielRosenwasser @RyanCavanaugh this is in the 3.6.1 milestone, but I don’t think it’s actionable without #24738, which, uh, probably cannot go into 3.6.

**Comment by thecotne** [comment]
@weswigham can you please explain how is this a bug?

it seems that if `copy` wants a mutable `Float32Array` and you give it an immutable (readonly) `Float32Array` it's defiantly a mistake because you can't modify immutable data but if you wanted mutable data then you probably will try to mutate it (if you have no plans to mutate data you can change type annotation to be immutable (readonly) like so `Readonly<Float32Array>` and it will work fine)
